geopandas/geodataframe.py

```diff
@@ -140,21 +212,61 @@ def set_geometry(self, col, drop=False, inplace=False, crs=None):
         if to_remove:
             del frame[to_remove]
 
+        if geo_column_name in frame.columns:
+            del frame[geo_column_name]
+
         if isinstance(level, GeoSeries) and level.crs != crs:
             # Avoids caching issues/crs sharing issues
             level = level.copy()
             level.crs = crs
 
-        # Check that we are using a listlike of geometries
-        if not all(isinstance(item, BaseGeometry) or not item for item in level):
-            raise TypeError("Input geometry column must contain valid geometry objects.")
-        frame[geo_column_name] = level
-        frame._geometry_column_name = geo_column_name
-        frame.crs = crs
-        frame._invalidate_sindex()
-        if not inplace:
+        if isinstance(level, Series):
+            level = GeoSeries(level)
+            level, _ = level.align(frame, join='right')
+        else:
+            level= GeoSeries(level, index=frame.index)
+
+        columns, index = frame._data.axes
+        blocks = frame._data.blocks
+
+        geom_block = level._data._block
+        geom_block = GeometryBlock(geom_block.values,
+                                   slice(len(columns), len(columns) + 1))
+        columns = columns.append(pd.Index([geo_column_name]))
+        blk_mgr = BlockManager(blocks + (geom_block,), [columns, index])
+
+        # frame[geo_column_name] = level
+        # frame._geometry_column_name = geo_column_name
+        # frame.crs = crs
+        # frame._invalidate_sindex()
+        if inplace:
+            frame._data = blk_mgr
+            frame._geometry_column_name = geo_column_name
+            frame.crs = crs
+            frame._invalidate_sindex()
+        else:
+            frame = self._constructor(blk_mgr, geometry=geo_column_name,
+                                      crs=crs)
             return frame
 
+    def _ensure_geometry(self):
+        return self
```

Comment:
Is this function a left-over of earlier tries? 
As I first thought it was not yet updated because it ignores the actual code impl below, but now I see this was also a new function (not an existing). Is a correct implementation of this still needed?